core/classifiers/classifier_utils.py

```diff
@@ -60,15 +59,40 @@ def extract_svm_parameters(clf):
     }
 
 
-def unicode_validator_for_classifier_data(var):
+def unicode_validator_for_classifier_data(classifier_data):
     """Validates that incoming object contains unicode literal strings."""
-    if isinstance(var, dict):
-        for k in var.keys():
+    if isinstance(classifier_data, dict):
+        for k in classifier_data.keys():
             if isinstance(k, str):
                 raise Exception('Expected %s to be unicode but found str.' % k)
-            unicode_validator_for_classifier_data(var[k])
-    if isinstance(var, (list, set, tuple)):
-        for item in var:
+            unicode_validator_for_classifier_data(classifier_data[k])
+    if isinstance(classifier_data, (list, set, tuple)):
+        for item in classifier_data:
             unicode_validator_for_classifier_data(item)
-    if isinstance(var, str):
-        raise Exception('Expected \'%s\' to be unicode but found str.' % var)
+    if isinstance(classifier_data, str):
+        raise Exception(
+            'Expected \'%s\' to be unicode but found str.' % classifier_data)
+
+
+def convert_float_numbers_to_string_in_classifier_data(classifier_data):
```

Comment:
Instead of this, can we do either of the following:
- update the dict to contain an "actual type" key that's float/floatlist, if the corresponding field is modified, or
- specify the set of fields per classifier type that should explicitly go through transformation, or
- put the fields that should go through transformation in a subdict that exclusively contains only fields that should go through transformation

I think it's better for the decryption mechanism to be deterministic in terms of what fields to decrypt, rather than "try converting all strings in the JSON to floats and see if that works". It's quite possible that in the future a string may convert fine but was actually supposed to be left as a string in the first place.

Reply:
TBH, I don't think we need a deterministic decryption mechanism. The reasons are:
* We can always use RegEx to determine whether a string is a floating point number or not. So that we don't have to use "try converting all strings" approach.
* Other than that, I don't think there ever would be a situation where the float number would actually be stored as a string. I can't think of anything that would require us to store float as string and not as float value.

Now, regarding the deterministic approach, I am not able to think of a completely deterministic approach. The main reason is the lists. Lists are heterogeneous in Python and so it can contain a float, a string and a dict. So we can't say that the actual type of list is float because it is applicable to it only partially. So even if we know that this list contains transformed values, we still have to use "try converting all strings to float and see if it works" approach. For singleton values, we can keep track of them by replacing the value with something like:
```
{
    'transformed_value': [transformed value],
    'original_type: 'type/float'
}
```

But as you see, it's the lists which make the approach only partially deterministic.

Regarding the 3 approaches that you mentioned, the approach I suggested above (for singleton values) closely matches your 1st approach. For other approaches:
* 2nd approach can't be applied because of recursive nature of the function. As you can see, we have no idea of what is the key when we are transforming a list. Its called recursively.
* 3rd approach can't be applied because the structure of the classifier data is equally important. Since a classifier data may contain a trained model for multiple classifiers which are applied hierarchically (SVM and KNN both). So by generating a subdict we lose information of the structure which is equal to saying that we don't have a trained classifier anymore.

So I suggest that we go through RegEx approach. As I mentioned previously, I don't see any reason of keeping float values as strings and not as floats.

Reply:
Hm, you have a good argument. I am not concerned about heterogeneous lists, since I don't think we'll have that in practice (it's generally not a great principle anyway and should be caught in code review since it means that the types of the elements within the lists are nondeterministic). I also feel that specifying all float locations seems quite tedious, especially since it looks like most of what's being sent is numbers anyway.

But ... I would still like a strict check on the conversions. It is at these seams between services that errors tend to occur, and these are hard to catch because of the lack of e2e tests that span both systems in a unified way.

So, here's another alternative, after looking at some of the sample output you sent me. Can we do the reverse, e.g. attach a list to the payload specifying which fields are actually strings? For example we could send an additional list alongside it that contains ['best_params.kernel', 'kernel_params.kernel']. Then the default would be to convert everything back to a float except the stuff explicitly keyed in this list, which we leave as-is.

Reply:
The thing is due to recursive nature of function it makes jt very much complex to prepare a list like that at top level of dict. However, it might be possible to have such list at each subdict level. So that each subdict contains inform regarding which of its attribute must not go through transformation.

Reply:
> So, here's another alternative, after looking at some of the sample output you sent me. Can we do the reverse, e.g. attach a list to the payload specifying which fields are actually strings? For example we could send an additional list alongside it that contains ['best_params.kernel', 'kernel_params.kernel']. Then the default would be to convert everything back to a float except the stuff explicitly keyed in this list, which we leave as-is.

I implemented a function to do this. PTAL. Also take a look at the corresponding test case in unit tests.